src/reader.rs

```diff
@@ -227,54 +207,121 @@ impl ReqwestReader {
 
 #[cfg(feature = "reqwest")]
 impl AsyncFileReader for ReqwestReader {
-    fn get_metadata_bytes(&self, range: Range<u64>) -> BoxFuture<'_, AsyncTiffResult<Bytes>> {
+    fn get_metadata_bytes(&mut self, range: Range<u64>) -> BoxFuture<'_, AsyncTiffResult<Bytes>> {
         self.make_range_request(range)
     }
 
-    fn get_image_bytes(&self, range: Range<u64>) -> BoxFuture<'_, AsyncTiffResult<Bytes>> {
+    fn get_image_bytes(&mut self, range: Range<u64>) -> BoxFuture<'_, AsyncTiffResult<Bytes>> {
         self.make_range_request(range)
     }
 }
 
 /// An AsyncFileReader that caches the first `prefetch` bytes of a file.
 #[derive(Debug)]
 pub struct PrefetchReader {
-    reader: Arc<dyn AsyncFileReader>,
-    buffer: Bytes,
+    reader: Box<dyn AsyncFileReader>,
+    /// Invariant: buffers are monotonically increasing buffers starting at the beginning of the
+    /// file
+    buffers: Vec<Bytes>,
+    /// The exponent used for deciding how much more data to fetch on overflow of the existing buffer.
+    ///
+    /// buffer_length ^ fetch_exponent
+    overflow_fetch_exponent: f64,
 }
 
 impl PrefetchReader {
     /// Construct a new PrefetchReader, catching the first `prefetch` bytes of the file.
-    pub async fn new(reader: Arc<dyn AsyncFileReader>, prefetch: u64) -> AsyncTiffResult<Self> {
+    pub async fn new(
+        mut reader: Box<dyn AsyncFileReader>,
+        prefetch: u64,
+        overflow_fetch_exponent: f64,
+    ) -> AsyncTiffResult<Self> {
         let buffer = reader.get_metadata_bytes(0..prefetch).await?;
-        Ok(Self { reader, buffer })
+        Ok(Self {
+            reader,
+            buffers: vec![buffer],
+            overflow_fetch_exponent,
+        })
+    }
+
+    /// Expand the length of buffers that have been pre-fetched
+    ///
+    /// Returns the desired range and adds it to the cached buffers.
+    async fn expand_prefetch(&mut self, range: Range<u64>) -> AsyncTiffResult<Bytes> {
+        let existing_buffer_length = self.buffer_length() as u64;
+        let additional_fetch =
+            (existing_buffer_length as f64).powf(self.overflow_fetch_exponent) as u64;
+
+        // Make sure that we fetch at least the entire desired range
+        let new_range =
+            existing_buffer_length..range.end.max(existing_buffer_length + additional_fetch);
+        let buffer = self.reader.get_metadata_bytes(new_range).await?;
+        self.buffers.push(buffer);
```

Comment:
This is where we request the overflow and add it to the existing buffering (`self.buffers.push(buffer)`)